keras/backend/mxnet_backend.py

```diff
@@ -2690,55 +2654,394 @@ def rnn(step_function, inputs, initial_states,
-        masks = list(mx.sym.split(mask.symbol, axis=1,
-                                  squeeze_axis=1, num_outputs=dshape[1]))
-        # Reverse the mask sequence
+    if unroll:
+        # Split the inputs across time dimension and generate the list of inputs
+        # with shape `(samples, ...)` (no time dimension)
+        inputs = list(mx.sym.split(inputs.symbol, axis=1,
+                                   squeeze_axis=1, num_outputs=dshape[1]))
+
+        # Reverse the input sequence
         if go_backwards:
-            masks.reverse()
+            inputs.reverse()
+
+        states = initial_states
+        outputs = []
+        prev_output = None
+
+        if mask is not None:
+            if not states:
+                raise ValueError('MXNet Backend: Initial states is not provided when masking is '
+                                 'enabled.')
+            if mask.dtype != dtype:
+                mask = cast(mask, dtype)
+            # Split the mask across time dimension and generate the list of masks
+            # with shape `(samples, 1)` (no time dimension)
+            masks = list(mx.sym.split(mask.symbol, axis=1,
+                                      squeeze_axis=1, num_outputs=dshape[1]))
+            # Reverse the mask sequence
+            if go_backwards:
+                masks.reverse()
+        else:
+            masks = [None for _ in inputs]
+
+        # Iterate over a time step
+        for inp, msk in zip(inputs, masks):
+            last_output, new_states = step_function(KerasSymbol(inp),
+                                                    states + constants)
+            if getattr(last_output, '_uses_learning_phase', False):
+                uses_learning_phase = True
+            if msk is not None:
+                new_states = [KerasSymbol(mx.sym.where(msk,
+                                                       ns.symbol,
+                                                       s.symbol))
+                              for s, ns in zip(states, new_states)]
+                # Initialize the output for first time step
+                if prev_output is None:
+                    prev_output = zeros_like(last_output)
+                last_output = KerasSymbol(mx.sym.where(msk,
+                                                       last_output.symbol,
+                                                       prev_output.symbol))
+                prev_output = last_output
+            states = new_states
+            # Expand the output dimension from `(samples, output_dim)` to
+            # `(samples, 1, output_dim)` with middle axis as time dimension
+            outputs.append(mx.sym.expand_dims(last_output.symbol, axis=1))
+        # Concatenate the output across time dimension
+        outputs = mx.sym.concat(*outputs, dim=1)
     else:
-        masks = [None for _ in inputs]
+        # TODO: remove version check after mxnet 1.3.1 release
+        if mx.__version__ != '1.3.1':
```

Comment:
This may not work always because you can have .post0 etc. releases. Please use regex for guaranteed checks